### DISCUSSION TITLE
Change prompt dynamically

### DISCUSSION THREAD

**Discussion by AnneVogt** [jspsych/jsPsych#1012] [Q&A] [answered]
Hello, 
I am asking participants to categorize a picture, using the image-keyboard-response plugin. The picture will be on screen for a fixed amount of time no matter when participants react. 
Furthermore, I use the prompt feature to display the two choice options (using keyboard buttons 'q' and 'p' respectively). To give participants the feeling that we are registering their reponse, I would like to change the prompt part depending on which button they pressed.

```var exercise_A = {
type: "image-keyboard-response",
stimulus: jsPsych.timelineVariable("stim"),
prompt: '<table><tr><tr><tr><td style="color:rgb(190,190,190);"> Vokal (Q)  </td><td></td><td></td><td style="color:rgb(190,190,190);"> Konsonant (P)</td></tr></table>',
choices: ["q", "p"],
trial_duration: 2000,
response_ends_trial: false,
data: jsPsych.timelineVariable("data"),
...
};
```
Ideally, 300ms before the trial ends, either "Vokal (Q)" or "Konsonant (P)" would turn blue (or later, if they respond late). Is this possible?

**Comment by becky-gilbert** [comment]
Disclaimer: there might be a better/simpler solution! But one way to do this is by adding IDs to the HTML elements that you might want to change ("Vokal (Q)" and "Konsonant (P)"), and by using the `on_load` function for the trial. You can use the trial's `on_load` function to:
(1) create a variable "prompt_id_to_change" that will store the ID of the element that you want to change, either "Vokal (Q)" or "Konsonant (P)", and initialize it to null
(2) add a 'keydown' event listener to the page, and inside that function, check to see which key was pressed, and change the value of your "prompt_id_to_change" variable accordingly
(3) use `setTimeout` to run a function when you want the feedback to be displayed - this function will check the "prompt_id_to_change" variable, and if it's not null, change the text color of the element with this ID.
Like this:
```javascript
var trial = {
            type: "image-keyboard-response",
            stimulus: jsPsych.timelineVariable("stim"),
            prompt: '<table><tr><tr><tr><td id="prompt-q" style="color:rgb(190,190,190);"> Vokal (Q)  </td><td></td><td></td><td id="prompt-p" style="color:rgb(190,190,190);"> Konsonant (P)</td></tr></table>',
            choices: ["q", "p"],
            trial_duration: 5000,
            stimulus_duration: 2000,
            response_ends_trial: false,
            on_load: function() {
                var feedback_time = 4500; // trial duration - 500 ms
                var prompt_id_to_change = null;
                function after_key_response(e) {
                    if (e.key == "q") {
                        prompt_id_to_change = "prompt-q"
                    } else if (e.key == "p") {
                        prompt_id_to_change = "prompt-p"
                    }
                }
                document.addEventListener('keydown',after_key_response);
                var feedback_timeout = setTimeout(function(){
                    if (prompt_id_to_change !== null) {
                        document.getElementById(prompt_id_to_change).style.color = "blue";
                    }
                }, feedback_time);
            }
        };
```
I set the feedback time to be 500 ms before the end of the trial because I think the trial loading takes a bit of time, so you need to account for that in the total trial duration.
Does this do what you want?

**Reply by AnneVogt** [reply]
Hi Becky, thanks for your suggestion. It seems to be what I want. However, when I run your code, nothing is changing - the trial is running normally, but the prompt colour does not change dynamically. Therefore, I wanted to print the variable "promp_id_to_change" to the console, to see if this value was correct. I added `console.log(prompt_id_to_change);`  just before the `document.addEventListener()` to my code but nothing is given out to the console so I am at loss how to find the source of the problem. 
Where should I put the `console.log()`-command? Which other options can I use to find out why the code it not producing the required result?

**Reply by becky-gilbert** [reply]
Hm, it worked for me so I'm not sure what the problem is. Would you mind posting your code?

**Reply by becky-gilbert** [reply]
You put a console.log inside of the trial's on_load function (but not inside another function or code block) and you didn't see anything print to the console? That sounds like the on_load function isn't running, which is weird. Could you double check by just putting `console.log('test')` in the on_load function on its own?

**Reply by AnneVogt** [reply]
It seems to be the case that the on_load function is not working. Putting console.log('test') inside the function doesn't lead to anything being printed to the console. However, putting a console.log-command inside the on_finish function lead to a result.

Here is a working example of this part of my experiment. I hope you will be able to help me tracking the error.

```<html>
<head>
<meta charset="utf-8">

<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>

<script src="jatos.js"></script>

<script src="jsPsych-master/jspsych.js"></script>
<script src="jsPsych-master/plugins/jspsych-call-function.js"></script>
<script src="jsPsych-master/plugins/jspsych-image-keyboard-response.js"></script>
<script src="jsPsych-master/plugins/jspsych-html-keyboard-response.js"></script>

 <link href="jsPsych-master/css/jspsych.css" rel="stylesheet"></link>
 </head>
<script>

var timeline=[];

/*Trial array for experiment*/
var item_pics_A = [
{item: 'stimuli/hose_r.png', block: 'BP_trial', stim_num: '1', data: {test_part: 'BP_trial', datei: 'hose_r.png', target_response: 'q', relatedness: 'r', distractor: 'Pyjama', bild: 'Hose', response_comp: 'comp', last_letter_pic: 'vow', ID: '1'}},
{item: 'stimuli/ei_r.png', block: 'BP_trial', stim_num: '2', data: {test_part: 'BP_trial', datei: 'ei_r.png', target_response: 'q', relatedness: 'r', distractor: 'Marmelade', bild: 'Ei', response_comp: 'comp', last_letter_pic: 'vow', ID: '2'}},
];

var fixation = {
	type: 'html-keyboard-response',
	stimulus: '<div style="font-size:60px;">+</div>',
	choices: jsPsych.NO_KEYS,
	trial_duration: 1000,
};

var inter_stimulus = {
	type: "html-keyboard-response",
	stimulus: "",
	trial_duration: 1000,
	choices: jsPsych.NO_KEYS,
	data: {test_part: "ITI"}
};

var trial = {
  type: "image-keyboard-response",
  stimulus: jsPsych.timelineVariable("item"),
  prompt: '<table><tr><tr><tr><td id="prompt-q" style="color:rgb(190,190,190);"> Vokal (Q)  </td><td></td><td></td><td id="prompt-p" style="color:rgb(190,190,190);"> Konsonant (P)</td></tr></table>',
  choices: ["q", "p"],
  trial_duration: 5000,
  stimulus_duration: 2000,
  response_ends_trial: false,
  on_load: function() {
    console.log('test');
      var feedback_time = 4500; // trial duration - 500 ms
      var prompt_id_to_change = null;
      function after_key_response(e) {
          if (e.key == "q") {
              prompt_id_to_change = "prompt-q"
          } else if (e.key == "p") {
              prompt_id_to_change = "prompt-p"
          }
      };
    console.log(prompt_id_to_change)
    document.addEventListener('keydown',after_key_response);
      var feedback_timeout = setTimeout(function(){
        if (prompt_id_to_change !== null) {
            document.getElementById(prompt_id_to_change).style.color = "blue";
        }
      }, );
  },
  data: jsPsych.timelineVariable("data"),
	on_finish: function(data){
  //console.log('finish')
	if(data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.target_response)){
	data.error_code = 1;
	} else {
	data.error_code = 0;
	}
  jsPsych.data.addDataToLastTrial(data);
	},
};

var trial_procedure_A = {
	timeline: [fixation, trial, inter_stimulus],
	timeline_variables: item_pics_A,
	randomize_order: true,
};

timeline.push(trial_procedure_A);

jsPsych.init({
timeline: timeline,
on_finish: function() {
var resultDataStr = jsPsych.data.get().csv();
var promise2 = jatos.uploadResultFile(resultDataStr, "example_PWI_BP.csv");
},
});

</script>

</html>
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @AnneVogt, thanks for sharing your code! 

Your on_load function works for me using jsPsych v6.1.0. I see that you're loading files from jsPsych-master rather than a specific release. It might be that there was a bug in the master branch that affected the on_load parameter for this plugin. Could you try testing your code with v6.1.0?

Using jsPsych 6.1.0 release, your code works for me with one key change: the `feedback_time` variable was missing as the second argument in your setTimeout function. Instead it should look like this:
```javascript
var feedback_timeout = setTimeout(function () {
  if (prompt_id_to_change !== null) {
    document.getElementById(prompt_id_to_change).style.color = "blue";
  }
}, feedback_time);
```
Let us know if that works!

**Reply by AnneVogt** [reply]
Thank you Becky for pointing out the missing variable `feedback_time` . The most important advice was to check my jsPsych-version. Indeed, the on_load-property had not been defined in the master branch. Unfortunately, simply downloading jsPsych v.6.1.0. and changing all the paths didn't work because many other plugins were not working properly anymore (resulting in many different types of errors, like plugin trial type not defined although the plugins were there). To keep my experiment running I copy-pasted the on_load-parts from the 6.1.0.-version into my installation. If you have suggestions how to circumvent this version problem and how to transfer experiments to the current version, I would be happy to know in order to avoid problems in the future.

**Reply by becky-gilbert** [reply]
Oh no, really sorry to hear about the problems switching to v6.1.0. I'm afraid I don't have any advice for how best to manage that problem. But I can sympathize because I once wrote a bunch of tasks using the master branch and had a similar thing happen! In the long run I think it's worth the effort to entirely switch from using a downloaded snapshot of the master branch to using a release, like v6.1.0, but that will mean having to resolve all of those errors. Good luck, and let us know if we can help with any specific problems.
